Implement boats

Postremus opened the pull request:
This pr attempts to implement boats, excluding:
- boat specific physics
 - Boats seem to be affected by gravity, and therefore sink when placed
 - Most of the deprecated API methods are not implemented

Some basic interactions, like damage calculation and sounds are also missing. They will be done in a future pr, since this one is already to big for my taste.

related to #504

SpaceManiac commented on src/main/java/net/glowstone/command/TitleCommand.java:
```diff
@@ -0,0 +1,233 @@
+     * @return the colored string, or null
+     */
+    // TODO: Replace with proper chat components when possible
+    public static String convertJson(JSONObject json) {
```
Judging by the docs, this can be replaced with `new TextMessage(json).flatten()`, which additionally supports non-color formatting.

dequis replied:
@turt2live seems like you forgot to address this one

SpaceManiac commented on src/main/java/net/glowstone/net/message/play/game/TitleMessage.java:
```diff
@@ -37,6 +41,26 @@ public TitleMessage(Action action) {
+        if (s == null) return "";
+
+        JSONObject parsed = TextMessage.convert(s);
+        if (parsed == null) return "";
```
`convert()` never returns null. `TextMessage.convert(s).toJSONString()` can be invoked through `new TextMessage(s).encode()` without having to additionally expose `convert`.

It might also make sense to make `text` a `TextMesssage` instead, which is the class acting as the placeholder for whatever eventually becomes the chat model.

turt2live commented:
@SpaceManiac This pull request is pending further review by yourself. Please take the time, if possible, to do so. Thanks!

turt2live commented:
Closing this PR - Project has moved on to a [new repo](https://github.com/GlowstoneMC/Glowstone). 

If this needs implementing in the newer repository I'll try and find time to implement it.

Postremus committed:
Implement boats

-implement storage for boats
-add steer boat message and handler implementation
-add Boat Items and spawn boats
-started implementation of vehiclemove handler

related to #504

Postremus committed:
Don't allow boats to move from water onto solid ground

Postremus committed:
spawn boats on top of water also

Postremus committed:
boats should burn in furnace

Postremus committed:
drop boat item on death

Postremus committed:
some documentation changes

aramperes commented on src/main/java/net/glowstone/net/handler/play/inv/VehicleMoveHandler.java:
```diff
@@ -1,12 +1,119 @@
+            VehicleMoveEvent event = EventFactory.callEvent(new VehicleMoveEvent((Vehicle) vehicle, oldLocation, newLocation));
+
+            if (!event.getTo().equals(newLocation)) {
+                vehicle.teleport(event.getTo(), TeleportCause.PLUGIN);
```
Is there a reason to set this to `TeleportCause.PLUGIN`?

Postremus replied:
Changed the implementation to be more in line with papers. Changing the to or from location of this event does not seem to do anything there.

Postremus committed:
vehiclemovement seems to just ignore changes to his location

aramperes commented:
Thank you for your contribution :)
